### ISSUE TITLE
__contains__ should narrow  Optional[T] to T 

### ISSUE THREAD

**Issue by roganov** [microsoft/pyright#802]
`__contains__` should narrow  Optional[T] to T 

**To Reproduce**
```
from typing import Optional
import random


def foo(p: str) -> None:
    print(p)


x: Optional[str]
if random.random() < 0.5:
    x = None
else:
    x = '2'

if x in ['2']:
    foo(x)
```

**Describe the bug**
`x` can not be `None` on the last line of this program, but error is reported:
```
  16:9 - error: Argument of type "str | None" cannot be assigned to parameter "p" of type "str" in function "foo"
  Type "None" cannot be assigned to type "str"
    Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
```

**Expected behavior**
No errors

**Comment by erictraut** [comment]
The `__contains__` method for the class `list` is defined as follows by builtins.pyi in typeshed:
```
class list(MutableSequence[_T], Generic[_T]):
    def __contains__(self, o: object) -> bool: ...
```
The `o` parameter is typed as `object`, so it's not restricted to `_T`. `None` is a valid `object`. And there's no way for the type checker to understand whether `list.__contains__` will return true or false for `None` based on the type information alone.

The type checker would need to include special-case knowledge about the internal implementation of the list class beyond what is provided in the type stub file, something which I have generally tried to avoid in Pyright.

I would normally say that this behavior is "as designed" and close the issue, but something gives me pause. It appears that mypy includes special-case logic to handle this case, so I'm going to think about this a little more. It may be worth adding this logic specifically for some of the common built-in classes (namely: list, tuple, set, frozenset, deque). I don't see any way to support it generally for all classes that provide support for the `__contains__` magic method.

In the meantime, here's a simple workaround you can use in your code. Just add the following line:
```
assert x is not None
```

**Comment by erictraut** [comment]
OK, I've implemented this in a limited form specifically for specialized objects of type list, set, frozenset, and deque. This support will be in the next version of Pyright.

**Comment by roganov** [comment]
Thank you.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.51, which I just published. It will be included in the next release of Pylance.
